### ISSUE TITLE
Dart 2 frontend uses the wrong scope for type annotations

### ISSUE THREAD

**Issue by rmacnak-google** [dart-lang/sdk#32752]
This is a simplification of an issue encountered with generated FIDL2 bindings in Fuchsia. It blocks enabling Dart 2 in Fuchsia.

```
import 'dart:math' as math;

class Foo {
  List<math.Point> math;
}

main() {
  print(new Foo());
}
```
works in Dart, but in Dart 2 yields

```
a.dart:4:8: Error: 'math.Point' can't be used as a type because 'math' doesn't refer to an import prefix.
  List<math.Point> math;
       ^^^^^^^^^^
```
@peter-ahe-google @a-siva @zanderso

**Comment by eernstg** [comment]
The behavior that you've seen in Dart 2 is correct, and the behavior that you've seen elsewhere is a bug.

Dart does not have a notion of putting declared names into different namespaces according to the nature of their declaration (with one exception: labels). So `math` in the type annotation is looked up in the current scope, which is the instance scope of `Foo`, and in that scope the result will be the instance variable. It is a bug for any Dart tool (Dart 1 as well as 2) to resolve that occurrence of `math` to the library prefix.

Taking that into account, wouldn't it be possible to generate code which doesn't have that name clash?

**Comment by kmillikin** [comment]
For what it's worth, this is a duplicate of https://github.com/dart-lang/sdk/issues/32200

**Comment by eernstg** [comment]
Also note #32248.

**Comment by zanderso** [comment]
@eernstg Despite possibly being a bug, this is long-standing behavior of the VM and apparently also the analyzer, so I think we should consider the new behavior to be a breaking change. Maybe we should have a Dart 2 breaking change announcement from the language team for this. /cc @leafpetersen @a-siva 

@rmacnak-google Do you think that mangling the generated library prefixes with a leading '$' would be sufficient? Can library prefixes have a leading '$'? /cc @abarth.

**Comment by eernstg** [comment]
Funny, the analyzer isn't even quite sure whether it's looking up `math` as a prefix:
```dart
import 'dart:math' as math;

class Foo {
  List<math.Point> math;
}

main() => print(new Foo());
```
produces
```
  hint • Unused import: 'dart:math' at n007.dart:4:8 • unused_import
```
I understand that even a bug fix can be a breaking change, organizationally. I hope it's manageable! On the other hand, I think we should maintain the property that "Dart has one name space" as far as possible (and nobody knows about labels, anyway ;-).

**Comment by leafpetersen** [comment]
Posting an announcement about this seems like a reasonable idea, I can do.

**Comment by vsmenon** [comment]
Is this still a P0?

**Comment by vsmenon** [comment]
Dropping the P0 as it appears @rmacnak-google unblocked Fuchsia.  Please bump if I misunderstood.

Is the only open here to communicate the change?

**Comment by leafpetersen** [comment]
I've announced that this is de facto changing.  We have a bug (https://github.com/dart-lang/sdk/issues/32248) for the analyzer to fix this, everything else gets this from the CFE.    Closing this, but re-open if there are any remaining issues that I've missed.
